util/ribbon_alg.h

```diff
@@ -0,0 +1,821 @@
+  struct NoopBacktrackStorage {
+    bool UseBacktrack() { return false; }
+    void BacktrackPut(Index, Index) {}
+    Index BacktrackGet(Index) {
+      assert(false);
+      return 0;
+    }
+  } nbts;
+  return BandingAddRange(ss, &nbts, bh, begin, end);
+}
+
+// ######################################################################
+// ######################### Solution Storage ###########################
+
+// Back-substitution and query algorithms unfortunately depend on some
+// details of data layout in the final data structure ("solution"). Thus,
+// there is no common SolutionStorage covering all the reasonable
+// possibilities.
+
+// ###################### SimpleSolutionStorage #########################
+
+// SimpleSolutionStorage is for a row-major storage, typically with no
+// unused bits in each ResultRow. This is mostly for demonstration
+// purposes as the simplest solution storage scheme. It is relatively slow
+// for filter queries.
+
+// concept SimpleSolutionStorage extends RibbonTypes {
+//   void PrepareForNumStarts(Index num_starts) const;
+//   Index GetNumStarts() const;
+//   ResultRow Load(Index slot_num) const;
+//   void Store(Index slot_num, ResultRow data);
+// };
+
+// Back-substitution for generating a solution from BandingStorage to
+// SimpleSolutionStorage.
+template <typename SimpleSolutionStorage, typename BandingStorage>
+void SimpleBackSubst(SimpleSolutionStorage *sss, const BandingStorage &ss) {
+  using CoeffRow = typename BandingStorage::CoeffRow;
+  using Index = typename BandingStorage::Index;
+  using ResultRow = typename BandingStorage::ResultRow;
+
+  constexpr auto kCoeffBits = static_cast<Index>(sizeof(CoeffRow) * 8U);
+  constexpr auto kResultBits = static_cast<Index>(sizeof(ResultRow) * 8U);
+
+  // A column-major buffer of the solution matrix, containing enough
+  // recently-computed solution data to compute the next solution row
+  // (based also on banding data).
+  std::array<CoeffRow, kResultBits> state;
+  state.fill(0);
+
+  const Index num_starts = ss.GetNumStarts();
+  sss->PrepareForNumStarts(num_starts);
+  const Index num_slots = num_starts + kCoeffBits - 1;
+
+  for (Index i = num_slots; i > 0;) {
+    --i;
+    CoeffRow cr = *const_cast<BandingStorage &>(ss).CoeffRowPtr(i);
+    ResultRow rr = *const_cast<BandingStorage &>(ss).ResultRowPtr(i);
+    // solution row
+    ResultRow sr = 0;
+    for (Index j = 0; j < kResultBits; ++j) {
+      // Compute next solution bit at row i, column j (see derivation below)
+      CoeffRow tmp = state[j] << 1;
+      int bit = BitParity(tmp & cr) ^ ((rr >> j) & 1);
+      tmp |= static_cast<CoeffRow>(bit);
+
+      // Now tmp is solution at column j from row i for next kCoeffBits
+      // more rows. Thus, for valid solution, the dot product of the
+      // solution column with the coefficient row has to equal the result
+      // at that column,
+      //   BitParity(tmp & cr) == ((rr >> j) & 1)
```

Comment:
Question: how to make sure `BitParity(tmp & cr) == ((rr >> j) & 1)` as `tmp` could be changed for next row?
For example, to simplify, assume `kResultBits = 1`, `r = 2`. then tmp only has 2 bits, the first bit `tmp[0]` (`tmp[x]` means the `x` bit in `tmp`) it makes sure: `(tmp[0]+cr[0])^(tmp[1]+cr[1]) == rr[0]`. But `tmp[1]` is not assigned yet and will be generated later. How do we make sure we don't need to go back and modify `tmp[0]`?

Reply:
Our initial back-substitution state is all zeros, though the initial state is unused because that would mean referencing an "out of bounds" slot, and unoccupied banding rows have cr as all zeros.

Each step through here we set new tmp[x+1] (bit x + 1) to old tmp[x] (bit x). Since we're moving backward through the rows, the second bit of tmp (aka tmp[1] or (tmp >> 1) & 1) will hold the value assigned to the "next" (but most recently assigned) row.

Reply:
I see, as the last `r` rows is an upper triangular matrix which will be used to set the initial state. I was confused that it is a band matrix.